### PR TITLE
OBLS-310 Backend for Pick up allocation

### PR DESCRIPTION
**Pull request by druchniewicz**

### CODE REVIEW

**Comment by druchniewicz** [comment]
Totally first initial version of allocation API.

I added support for AUTO/MANUAL modes.
In AUTO mode I just create pick list based on order item (without distinction between display/warehouse pick).
in MANUAL mode I create pick list based on array of allocations where one object contains inventoryItemId, binLocationId, quantity).

File: grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        if (mode.equalsIgnoreCase(AllocationType.AUTO.name())) {
+            stockMovementService.createPicklist(orderItem, false)
+        } else if (mode.equalsIgnoreCase(AllocationType.MANUAL.name())) {
+            List<ItemToAllocate> allocations = data.allocations as List<ItemToAllocate>
+            allocations.each { itemToAllocate ->
+                stockMovementService.createOrUpdatePicklistItem(
+                        orderItem.requisitionItem,
+                        null,
+                        InventoryItem.get(itemToAllocate.inventoryItemId),
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: Should use load() here so there's no database query. However, the only bummer is that we won't know if the record exists or not until we save the picklist item (which isn't a big deal as long as the validation for the picklist item works properly - i.e. fails if fks are null).

File: grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    void allocate(StockMovementItem orderItem, Map data = [:]) {
+        String mode = data.mode as AllocationType
+        if (!mode) {
+            throw new IllegalStateException("mode parameter not provided")
+        }
+
+        if (mode.equalsIgnoreCase(AllocationType.AUTO.name())) {
+            stockMovementService.createPicklist(orderItem, false)
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: I assume this creates a picklist if one does not exist? 


@awalkowiak The one thing we haven't discussed / designed is whether the requisition / picklist state machine can handle the fact that items might not be allocated yet. 

In other words, we usually move all items from state to state at the same time. The pickup allocation flow changes that as we'll have some items that have been allocated and others that have not. So there are going to be gaps in multiple state machine that require us to be somewhat defensive about moving ALL items from state to state. 

An example would be a pick up order like this. We cannot move this order to "complete" until all items are either fulfilled (which probably means each item has been "staged" but i need to double check the requirements). 

- Requisition Item 1
    - Status: Staged
    - Requested: 10 
    - Allocated: 10
    - Picked: 10
    - Staged: 10
- Requisition Item 2
    - Status: Requested
    - Requested: 5 
    - Allocated: 0
    - Picked: 0
    - Staged: 0
    - Canceled: 0
- Requisition Item 3
    - Status: Canceled due to stock out
    - Requested: 2 
    - Allocated: 0
    - Picked: 0
    - Staged: 0
    - Canceled: 2

File: grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        } else if (mode.equalsIgnoreCase(AllocationType.MANUAL.name())) {
+            List<ItemToAllocate> allocations = data.allocations as List<ItemToAllocate>
+            allocations.each { itemToAllocate ->
+                stockMovementService.createOrUpdatePicklistItem(
+                        orderItem.requisitionItem,
+                        null,
+                        InventoryItem.get(itemToAllocate.inventoryItemId),
+                        Location.get(itemToAllocate.binLocationId),
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: Same

File: grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+        outboundOrder.lineItems.each { StockMovementItem item ->
+            item.availableItems = stockMovementService.getAvailableItems(outboundOrder.origin, RequisitionItem.get(item.id), false)
+        }
+
+        return outboundOrder
+    }
+
+    void allocate(StockMovementItem orderItem, Map data = [:]) {
+        String mode = data.mode as AllocationType
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: This should be 
```
AllocationType mode = data.mode as AllocationType
```

Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: And what is AllocationType? Did we define that previously? I think I called this AllocationMode in the design, so let's go with that.

**Review by jmiranda — COMMENTED** [review]
The was a first pass code review from yesterday. Just got around to posting it. Sorry!

**Comment by jmiranda** [comment]
> Totally first initial version of allocation API.

This looks great 

> (without distinction between display/warehouse pick).

That's fine for now. Getting allocation working in general is the key here so we can move forward with picking and staging. When you get a change, please create a ticket to add a separate AllocationService that calls an AllocationStrategy (FEFO, DISPLAY_FIRST, WAREHOUSE_FIRST, etc)? I'll probably need to add some additional specification to that ticket before you can get started but feel free to implement the display vs warehouse as you see fit.

File: grails-app/domain/org/pih/warehouse/requisition/RequisitionItem.groovy [hunk 5]
```diff
@@ -635,6 +635,30 @@ class RequisitionItem implements Comparable<RequisitionItem>, Serializable {
...
+
+    Boolean isPartiallyAllocated() {
+        def quantityAllocated = calculateQuantityAllocated() ?: 0
+        def quantityRequired = calculateQuantityRequired() ?: 0
+
+        return quantityAllocated > 0 && quantityAllocated < quantityRequired
+    }
+
+    String getAllocationStatus() {
+        if (isAllocated()) return "ALLOCATED"
```
Comment (druchniewicz) [inline grails-app/domain/org/pih/warehouse/requisition/RequisitionItem.groovy]: consider any enum for these statuses? or extend the existing RequistionItemStatus enum

File: grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    StockMovement get(String id) {
+        StockMovement outboundOrder = stockMovementService.getStockMovement(id)
+
+        if (!outboundOrder) {
+            throw new IllegalArgumentException("No outbound order found for id ${params.id}")
+        }
+
+        outboundOrder.lineItems.each { StockMovementItem item ->
+            item.availableItems = stockMovementService.getAvailableItems(outboundOrder.origin, RequisitionItem.get(item.id), false)
+            item.allocationStatus = RequisitionItem.get(item.id).getAllocationStatus()
```
Comment (druchniewicz) [inline grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: this probably can be refactored i.e. add new fields to StockMovementItem view domain class and set these fields during StockMovementItem creation (from RequisitionItem)

File: grails-app/domain/org/pih/warehouse/requisition/RequisitionItem.groovy (hunk 5 above)
Comment (jmiranda) [reply grails-app/domain/org/pih/warehouse/requisition/RequisitionItem.groovy]: yeah an AllocationStatus enum sounds good

File: grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy (hunk 6 above)
Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: +1

**Review by jmiranda — COMMENTED** [review]
I just realized my comments weren't posted yet. Sorry about that. I added these awhile ago.

File: grails-app/controllers/org/pih/warehouse/outboundOrder/OutboundOrderApiController.groovy [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+        StockMovementItem orderItem = outboundOrder.lineItems.find { StockMovementItem lineItem ->
+            lineItem.id == params.itemId
+        }
+
+        if (!orderItem) {
+            throw new IllegalArgumentException("Order item with id ${params.itemId} not found")
+        }
+
+        outboundOrderService.allocate(orderItem, jsonBody)
```
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/outboundOrder/OutboundOrderApiController.groovy]: I think we should just call this AllocationService. And I don't love the jsonBody being passed here. Let's make the arguments explicits or use the AllocationRequest class to pass in all of the properties need to make a decision.
```
AllocationResult allocate(AllocationRequest request)
```
The AllocationService should eventually delegate to allocation strategies as described in the PUA ticket. I don't think we necessarily need that just yet, but wanted to explain what will eventually happen.

Alos, the AllocationResult is a list of allocations. But we might just be call these Allocation or AllocationItems (instead of ItemToAllocate)

File: grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    void allocate(StockMovementItem orderItem, Map data = [:]) {
+        String mode = data.mode as AllocationType
+        if (!mode) {
+            throw new IllegalStateException("mode parameter not provided")
+        }
+
+        if (mode.equalsIgnoreCase(AllocationType.AUTO.name())) {
+            stockMovementService.createPicklist(orderItem, false)
+        } else if (mode.equalsIgnoreCase(AllocationType.MANUAL.name())) {
+            List<ItemToAllocate> allocations = data.allocations as List<ItemToAllocate>
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: yeah let's just call it an Allocation for now

File: grails-app/services/org/pih/warehouse/allocation/AllocationService.groovy [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    StockMovement getOutboundOrder(String id) {
+        StockMovement outboundOrder = stockMovementService.getStockMovement(id)
+
+        if (!outboundOrder) {
+            throw new IllegalArgumentException("No outbound order found for id ${id}")
+        }
+
+        outboundOrder.lineItems.each { StockMovementItem item ->
+            RequisitionItem requisitionItem = RequisitionItem.get(item.id)
+            item.availableItems = stockMovementService.getAvailableItems(outboundOrder.origin, requisitionItem, false)
```
Comment (druchniewicz) [inline grails-app/services/org/pih/warehouse/allocation/AllocationService.groovy:29]: I'm fetching outboundOrder (StockMovement) first and then adding all required data to particular line item like availableItems, allocations, allocationStatus, quantityAllocated. Thanks to this when I trigger fetch order endpoint I have all required data which I need to display in the mobile application and I don't need to make multiple requests to fetch all needed data

File: grails-app/services/org/pih/warehouse/outboundOrder/AllocationService.groovy [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            }
+
+            item.allocationStatus = requisitionItem.getAllocationStatus()
+            item.quantityAllocated = requisitionItem.calculateQuantityAllocated()
+        }
+
+        return outboundOrder
+    }
+
+    AllocationDetailsDto allocate(String requisitionItemId, AllocationType mode, List<AllocationDto> allocations) {
```
Comment (druchniewicz) [inline grails-app/services/org/pih/warehouse/outboundOrder/AllocationService.groovy]: it supports create and update. AllocationDto contains "id" field (PicklistItem id). When this id is sent by the mobile app it looks for existing picklist item, finds and updates it, otherwise picklistItem = null is sent to createOrUpdatePicklistItem method and new pick list item is created

File: grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy (hunk 6 above)
Comment (druchniewicz) [reply grails-app/services/org/pih/warehouse/outboundOrder/OutboundOrderService.groovy]: hmm, probably it will not be that trivial as I initially thought. Data in RequisitionItem like quantity allocated, allocation status etc. are also calculated on the fly, they are not stored in Requisition Item entity so in stock-movement-item.sql I cannot just use "select requisition_item.quantity_allocated" etc.
I leave it as it is for now

**Comment by druchniewicz** [comment]
@jmiranda @awalkowiak can you take a look at PR? I assume that can be much remarks requested from you to change but I tried to implement +/- first working version of PUA (backend + frontend). I shared the video on slack how it currently works and looks. You can take a look. I didn't test any special/edge cases etc.

File: grails-app/domain/org/pih/warehouse/requisition/RequisitionItem.groovy [hunk 11]
```diff
@@ -635,6 +636,30 @@ class RequisitionItem implements Comparable<RequisitionItem>, Serializable {
         return numInventoryItem
     }
 
+    def calculateQuantityAllocated() {
+        return PicklistItem.findAllByRequisitionItem(this).sum { it.quantity }
```
Comment (awalkowiak) [inline grails-app/domain/org/pih/warehouse/requisition/RequisitionItem.groovy]: I think this should also take into account substitutions - the same way as it is done in the `calculateQuantityPicked`.

File: src/main/groovy/org/pih/warehouse/allocation/AllocationStrategy.java [hunk 12]
```diff
@@ -0,0 +1,5 @@
+package org.pih.warehouse.outboundOrder;
+
+public enum AllocationStrategy {
```
Comment (awalkowiak) [inline src/main/groovy/org/pih/warehouse/allocation/AllocationStrategy.java:3]: I think it should rather belong to the Allocation domain, but we could refactor that later

**Review by awalkowiak — COMMENTED** [review]
Overall it looks good to me, however I would move out parts that are Allocation related to the Allocation domain related package instead having them in OutboundOrder (AllocationDto, AllocationStatus, AllocationStrategy, etc)

File: grails-app/domain/org/pih/warehouse/requisition/RequisitionItem.groovy (hunk 11 above)
Comment (druchniewicz) [reply grails-app/domain/org/pih/warehouse/requisition/RequisitionItem.groovy]: done

File: src/main/groovy/org/pih/warehouse/allocation/AllocationStrategy.java (hunk 12 above)
Comment (druchniewicz) [reply src/main/groovy/org/pih/warehouse/allocation/AllocationStrategy.java:3]: I move all to the allocation packages (not to outboundOrder packages)

**Comment by druchniewicz** [comment]
> Overall it looks good to me, however I would move out parts that are Allocation related to the Allocation domain related package instead having them in OutboundOrder (AllocationDto, AllocationStatus, AllocationStrategy, etc)

I move all to the allocation packages (not to outboundOrder packages)

File: grails-app/services/org/pih/warehouse/allocation/AllocationService.groovy (hunk 9 above)
Comment (jmiranda) [reply grails-app/services/org/pih/warehouse/allocation/AllocationService.groovy:29]: I think that's ok for now, but sorry my point was to move the allocation work to the AllocationService. The fetching of data could still be done through the OutboundOrderService.

We can work on refactoring this later. Let's just try to get PUA finished so we can demo and UAT as soon as possible.

File: src/main/groovy/org/pih/warehouse/api/StockMovementItem.groovy [hunk 13]
```diff
@@ -184,7 +193,10 @@ class StockMovementItem {
                         quantity      : inventoryItem.quantity
                 ],
                 manuallyAdded          : manuallyAdded,
-                availableItems         : availableItems
+                availableItems         : availableItems,
+                allocations            : allocations,
+                quantityAllocated      : quantityAllocated,
+                allocationStatus       : allocationStatus
```
Comment (jmiranda) [inline src/main/groovy/org/pih/warehouse/api/StockMovementItem.groovy:199]: This is ok for now, but we need to be careful adding new associations to the StockMovementItem because this is used on pages that need a list of 400+ stock movement items. Therefore, it might become a bit of a performance issue, particularly if these are generating additional N+1 queries. 

Again, keep it for now, but let's try to create new DTO classes for new APIs so we can start to untangle some of the problems with the Stock Movement API. If someone creates a ticket for this, I'll propose an resource designs for these APIs.

File: src/main/groovy/org/pih/warehouse/allocation/AllocationDetailsDto.groovy [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package org.pih.warehouse.allocation
+
+class AllocationDetailsDto {
```
Comment (jmiranda) [inline src/main/groovy/org/pih/warehouse/allocation/AllocationDetailsDto.groovy:3]: In the future, let's drop the Dto suffix. AllocationDetails is sufficient.


I could be wrong about this so feel free to argue. But I feel like if we design the class hierarchy intelligently enough there's no reason to make a distinction between a domain (, a DTO (returned by APIs), and/or a POJO (used locally).

File: src/main/groovy/org/pih/warehouse/allocation/AllocationDetailsDto.groovy [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+    String requisitionItemId
```
Comment (jmiranda) [inline src/main/groovy/org/pih/warehouse/allocation/AllocationDetailsDto.groovy:4]: In the future we should probably just make this Product unless there's a compelling reason to include the RequisitionItem. The method is attempting to allocate stock for the Product, not the RequisitionItem.

File: grails-app/services/org/pih/warehouse/allocation/AllocationService.groovy [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+            }
+
+            item.allocationStatus = requisitionItem.getAllocationStatus()
+            item.quantityAllocated = requisitionItem.calculateQuantityAllocated()
+        }
+
+        return outboundOrder
+    }
+
+    AllocationDetailsDto allocate(String requisitionItemId, AllocationMode mode, List<AllocationDto> allocations, List<AllocationStrategy> strategies = []) {
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/allocation/AllocationService.groovy:48]: For the future: The allocate() method should just be returning suggested allocations based on the strategy or strategies for the given product. Creating the picklist items can happen once the allocation has been executed for the entire order or order item. This would allow us to perform a dry-run allocation to show the allocation suggestions.

File: grails-app/services/org/pih/warehouse/allocation/AllocationService.groovy [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+        RequisitionItem requisitionItem = RequisitionItem.get(requisitionItemId)
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/allocation/AllocationService.groovy:49]: I think it's ok to pass the RequisitionItem for now, but if we separate the allocation from picklist creation, do we need requisition item? Would product suffice?

**Review by jmiranda — APPROVED** [review]
Again, this looks good to me. Discuss with Artur to see if he's ok with merging. If possible it would be great to merge on Tuesday so I can showcase on the standup. 

Note: @druchniewicz  I think you mentioned over Slack that I could download the two PR branches (frontend and backend) to perform the showcase, but that's going to take a lot of time for me and I'm already spread thin. So I'd rather wait until the frontend is ready, built, and available in appetize.io and the backend has been deployed to vvg.openboxes.com.

**Comment by druchniewicz** [comment]
> Again, this looks good to me. Discuss with Artur to see if he's ok with merging. If possible it would be great to merge on Tuesday so I can showcase on the standup.
> 
> Note: @druchniewicz I think you mentioned over Slack that I could download the two PR branches (frontend and backend) to perform the showcase, but that's going to take a lot of time for me and I'm already spread thin. So I'd rather wait until the frontend is ready, built, and available in appetize.io and the backend has been deployed to vvg.openboxes.com.

@awalkowiak if you're ok with merging it then let's merge it (but frontend needs to be merged as well then). Then all everything will be easy to run (backend will be automatically deployed on vvg.openboxesc.com) and mobile app will be deployed automatically as well.

As we discussed on slack, probably better will be to do some basic things work, make improvements/code refactors later? Up to you, but as we talked, smaller PRs = easier PRs = open for shorter time = probably more frequent feedback

**Comment by druchniewicz** [comment]
> Again, this looks good to me. Discuss with Artur to see if he's ok with merging. If possible it would be great to merge on Tuesday so I can showcase on the standup.
> 
> Note: @druchniewicz I think you mentioned over Slack that I could download the two PR branches (frontend and backend) to perform the showcase, but that's going to take a lot of time for me and I'm already spread thin. So I'd rather wait until the frontend is ready, built, and available in appetize.io and the backend has been deployed to vvg.openboxes.com.

Right, it would be time-consuming action. Anyway both PRs (backend, mobile app) have been already merged.
